dialogue-core/src/test/java/com/palantir/dialogue/core/PinUntilErrorChannelTest.java

```diff
@@ -0,0 +1,156 @@
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.palantir.dialogue.Response;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class PinUntilErrorChannelTest {
+
+    private LimitedChannel channel1 = mock(LimitedChannel.class);
+    private LimitedChannel channel2 = mock(LimitedChannel.class);
+    private Ticker clock = mock(Ticker.class);
+
+    private ImmutableList<LimitedChannel> channels = ImmutableList.of(channel1, channel2);
+    private PinUntilErrorChannel pinUntilErrorWithoutReshuffle;
+    private PinUntilErrorChannel pinUntilError;
+
+    @Before
+    public void before() {
+        pinUntilErrorWithoutReshuffle =
+                new PinUntilErrorChannel(new PinUntilErrorChannel.ConstantNodeList(channels, new Random(12345L)));
+        pinUntilError = new PinUntilErrorChannel(
+                new PinUntilErrorChannel.ReshufflingNodeList(channels, new Random(12893712L), clock));
+    }
+
+    @Test
+    public void channels_are_shuffled_initially_and_successful_requests_stay_on_channel() {
+        setResponse(channel1, 200);
+        setResponse(channel2, 204);
+
+        assertThat(getCode(pinUntilErrorWithoutReshuffle)).isEqualTo(204);
+        assertThat(getCode(pinUntilErrorWithoutReshuffle)).isEqualTo(204);
+        assertThat(getCode(pinUntilErrorWithoutReshuffle)).isEqualTo(204);
+    }
+
+    @Test
+    public void various_error_status_codes_cause_node_switch() {
+        for (int errorStatus = 300; errorStatus < 600; errorStatus++) {
+            before();
+            setResponse(channel1, 100);
+            setResponse(channel2, 204);
+
+            assertThat(IntStream.range(0, 6).map(number -> getCode(pinUntilErrorWithoutReshuffle)))
+                    .describedAs("Should be locked on to channel2 initially")
+                    .contains(204, 204, 204, 204, 204, 204);
+
+            setResponse(channel2, errorStatus);
+
+            assertThat(IntStream.range(0, 6).map(number -> getCode(pinUntilErrorWithoutReshuffle)))
+                    .describedAs("A single error code should switch us to channel 1")
+                    .contains(errorStatus, 100, 100, 100, 100, 100);
+        }
+    }
+
+    @Test
+    public void reshuffle_happens_roughly_every_10_mins() {
+        setResponse(channel1, 100);
+        setResponse(channel2, 204);
+
+        assertThat(IntStream.range(0, 6).map(number -> getCode(pinUntilError)))
+                .describedAs("First batch on channel1")
+                .contains(100, 100, 100, 100, 100, 100);
+
+        when(clock.read()).thenReturn(Duration.ofMinutes(11).toNanos());
+        assertThat(IntStream.range(0, 6).map(number -> getCode(pinUntilError)))
```

Comment:
this might be flaky since reshuffle could result in the same order as before no?

Reply:
I seed the Random instance at the beginning of the test, so it's 100% deterministic.